### ISSUE TITLE
Runtime error while calculating gradients , Version 1.5.0 specific

### ISSUE THREAD

**Issue by HarmanDotpy** [pytorch/pytorch#37573]
## 🐛 Bug

I wrote my code on google colab and pytorch was at version just before 1.5 then, after colab automatically shifted to version 1.5.0 of pytorch , my code started to give the following error while calculating the gradients
## To Reproduce
![image](https://user-images.githubusercontent.com/50492433/80685786-55cf1500-8ae5-11ea-8df0-b7cf8918dbc7.png)

## Expected behavior

There shouldn't be an error if there is a change of version, as I feel I am not doing any in place operation in the code, otherwise it would have given an error even for version before this, which is not the case.
The error is only for version 1.5.0

## Environment - GOOGLE COLAB

PyTorch version: 1.5.0+cu101
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.3
[pip3] torch==1.5.0+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.6.0+cu101
[conda] Could not collect
## Additional context

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by HapeMask** [comment]
I was just about to post an issue for what I think is the same problem, so I'll share my repro here instead:

```python
import torch as t
import numpy as np

th.autograd.set_detect_anomaly(True)

x = th.nn.Parameter(th.randn(3))
y = th.nn.Parameter(th.randn(3))
z = th.nn.Parameter(th.randn(3))

optim_a = th.optim.Adam([x, y], 1e-3)
optim_b = th.optim.Adam([y, z], 1e-3)

loss_a = ((x+y)**2).sum()

loss_b = (y**2).sum()
#loss_b = ((y+z)**2).sum() # Oddly, if you compute this instead, there is no inplace error.

loss_a.backward(retain_graph=True)
optim_a.step()

# If you move the computation of `loss_b` down here instead, there is also no inplace error.
loss_b.backward()
```

I can confirm this issue was not present in 1.4 or in a few nightlies between 1.4.0 and 1.5.0.

**Comment by albanD** [comment]
Hi,

Yes this is a bug we fixed. You can check the "Backwards Incompatible Changes" section of the 1.5 release doc [here](https://github.com/pytorch/pytorch/releases). In particular, the `torch.optim` section in there.
As discussed there, we fixed the inplace modification tests for the optimizers using new feature we have (removing `.data`). So this most likely means that the gradients computed previously were wrong :/

Let us know if you have any other question!

**Comment by HarmanDotpy** [comment]
Hi @albanD 
Thanks for your prompt reply on the issue.
Could you please explain a bit more on how the gradients were calculated earlier and what mistake you feel was there while calculating the gradients in previous versions of PyTorch.
This was something I could not understand.

Thanks

**Comment by albanD** [comment]
For the simple code sample just above, the error was that the optimizer.step() function was using `.data` which was hiding to the autograd the fact that it modifies the weights inplace.
But the original value of the weights are needed to perform the second backward and so with old versions, the second backward will compute wrong gradients.
The new version properly detects that the weights have been modified and so fails to compute the second backward as the value it needs is not available anymore.

**Comment by HarmanDotpy** [comment]
Oh, I see, that's why its a problem specific to 'retail graph = True' case. Thanks a lot.
